pkg/controller/pravegacluster/upgrade.go

```diff
@@ -404,6 +405,139 @@ func (r *ReconcilePravegaCluster) syncSegmentStoreVersion(p *pravegav1alpha1.Pra
+	if util.IsClusterUpgradingTo07(p) {
+		name = util.StatefulSetNameForSegmentstoreBelow07(p.Name)
+	} else {
+		name = util.StatefulSetNameForSegmentstoreAbove07(p.Name)
+	}
+
+	err = r.client.Get(context.TODO(), types.NamespacedName{Name: name, Namespace: p.Namespace}, oldsts)
+	//this check is to see if the old sts is present or not
+	if err != nil && errors.IsNotFound(err) {
+		err = r.client.Get(context.TODO(), types.NamespacedName{Name: newsts.Name, Namespace: p.Namespace}, newsts)
+		//this checks is to verify if the old sts is not present new sts should be present
+		if err != nil && errors.IsNotFound(err) {
+			return false, err
+		}
+		//this is the condition where we are checking if the new ready ss pods are equal to the ss replicas thus meaning upgrade is completed
+		if newsts.Status.ReadyReplicas == p.Spec.Pravega.SegmentStoreReplicas {
+			return true, nil
+		}
+	}
+
+	//this is to check if some pod has come into an error state or not
+	if oldsts.Status.ReadyReplicas+newsts.Status.ReadyReplicas != p.Spec.Pravega.SegmentStoreReplicas {
+		//this will get all the pods created with this target version till now
+		pods, err := r.getStsPodsWithVersion(newsts, p.Status.TargetVersion)
+		if err != nil {
+			return false, err
+		}
+		//checking if any of above pods have gone into error sate
+		_, err = r.checkUpdatedPods(pods, p.Status.TargetVersion)
+		if err != nil {
+			// Abort if there is any errors with the updated pods
+			return false, fmt.Errorf("updating statefulset (%s) failed due to %v", newsts.Name, err)
+		}
+	}
+
+	p.Status.UpdateProgress(pravegav1alpha1.UpdatingSegmentstoreReason, "0")
+	//this check to ensure that the oldsts always decrease by 2 as well as newsts pods increase by 2 only then the next increment or decrement happen
+	//first condition is true and used in the case of rollbackCondition
+	//second condition is true and used in case of Upgrade
+	if (r.IsClusterRollbackingFrom07(p) && newsts.Status.ReadyReplicas == *newsts.Spec.Replicas) || (oldsts.Status.ReadyReplicas+newsts.Status.ReadyReplicas == p.Spec.Pravega.SegmentStoreReplicas && newsts.Status.ReadyReplicas == *newsts.Spec.Replicas) {
+		//this check is run till the value of old sts replicas is greater than 0 and will increase two replicas of the new sts and delete 2 replicas of the old sts
+		if *oldsts.Spec.Replicas > 2 {
+			*newsts.Spec.Replicas = *newsts.Spec.Replicas + 2
+			err = r.client.Update(context.TODO(), newsts)
+			if err != nil {
+				return false, fmt.Errorf("updating statefulset (%s) failed due to %v", newsts.Name, err)
+			}
+			*oldsts.Spec.Replicas = *oldsts.Spec.Replicas - 2
+			err = r.client.Update(context.TODO(), oldsts)
+			if err != nil {
+				return false, fmt.Errorf("updating statefulset (%s) failed due to %v", oldsts.Name, err)
+			}
+		} else {
+			//here we remove the pvc's attached with the old sts and deleted it when old sts replicas have become 0
+			*newsts.Spec.Replicas = p.Spec.Pravega.SegmentStoreReplicas
```

Comment:
Again encapsulate changes to old STS in a method and invoke the same....
One more thing, we need to increment the newSTS first and then do stuff for deleting the old sts, so that we err on side of caution.... and can have more redundancy than less....

Reply:
done